pkg/controller/machinemanagement/machinemanagement_controller.go

```diff
@@ -0,0 +1,263 @@
+	// Return early if cluster deployment was deleted
+	if !cd.DeletionTimestamp.IsZero() {
+		if controllerutil.ContainsFinalizer(cd, hivev1.FinalizerMachineManagementTargetNamespace) {
+			if cd.Spec.MachineManagement.TargetNamespace != "" {
+				// Clean up namespace
+				cdLog.Info("Deleting target namespace ", cd.Spec.MachineManagement.TargetNamespace)
+				ns := &corev1.Namespace{
+					ObjectMeta: metav1.ObjectMeta{Name: cd.Spec.MachineManagement.TargetNamespace},
+				}
+				if err := r.Delete(context.TODO(), ns); err != nil && !apierrors.IsNotFound(err) {
+					return reconcile.Result{}, fmt.Errorf("failed to delete namespace: %w", err)
+				}
+			}
+			// Remove finalizer from cluster deployment
+			controllerutil.RemoveFinalizer(cd, hivev1.FinalizerMachineManagementTargetNamespace)
+			if err := r.Update(context.TODO(), cd); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to remove finalizer from cluster deployment: %w", err)
+			}
+		}
+		return reconcile.Result{}, nil
+	}
+
+	if cd.Spec.MachineManagement != nil && cd.Spec.MachineManagement.Central != nil {
+		targetNamespace := cd.Spec.MachineManagement.TargetNamespace
+		if targetNamespace == "" {
+			targetNamespace = apihelpers.GetResourceName(cd.Name+"-targetns", utilrand.String(5))
+		}
+
+		if cd.Spec.MachineManagement.TargetNamespace == "" {
+			cd.Spec.MachineManagement.TargetNamespace = targetNamespace
+
+			// Ensure the cluster deployment has a finalizer for cleanup
+			if !controllerutil.ContainsFinalizer(cd, hivev1.FinalizerMachineManagementTargetNamespace) {
+				controllerutil.AddFinalizer(cd, hivev1.FinalizerMachineManagementTargetNamespace)
+			}
+
+			if err := r.Update(context.TODO(), cd); err != nil {
+				cdLog.WithError(err).Log(controllerutils.LogLevel(err), "failed to update cluster deployment")
+				return reconcile.Result{Requeue: true}, nil
+			}
+		}
+
+		ns := &corev1.Namespace{}
+		if err := r.Get(context.Background(), types.NamespacedName{Name: targetNamespace}, ns); err != nil && apierrors.IsNotFound(err) {
+			cdLog.Info("Creating the target namespace ", targetNamespace)
+			ns.Name = targetNamespace
+			if err := r.Create(context.TODO(), ns); err != nil && !apierrors.IsAlreadyExists(err) {
+				return reconcile.Result{}, fmt.Errorf("failed to create target namespace %q: %w", targetNamespace, err)
+			}
+		}
+
+		// Ensure targetNamespace has machine management annotation
+		if err := r.addAnnotationToTargetNamespace(cd, cdLog, targetNamespace); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		credentialsSecretName := utils.CredentialsSecretName(cd)
+		credentialsSecret := &corev1.Secret{}
+		if err := r.Get(context.Background(), types.NamespacedName{Name: credentialsSecretName, Namespace: targetNamespace}, credentialsSecret); err != nil && apierrors.IsNotFound(err) {
```

Comment:
Can you add a TODO comment that we need to keep this in sync if it changes in the source, and file a Jira in this sprint? That would need a watch, and a comparison and update.

Reply:
https://issues.redhat.com/browse/HIVE-1412